### ISSUE TITLE
Not expanding unquoted variables correctly

### ISSUE THREAD

**Issue by pdpwebsecurify** [ljharb/shell-quote#1]
Consider the following example:

```bash
$ export T="c d"
$ node -e 'console.log(JSON.stringify(process.env.T), process.argv)' a b $T
```

You get the following result:

```js
"c d" [
  '/home/ec2-user/.nvm/versions/node/v17.0.1/bin/node',
  'a',
  'b',
  'c',
  'd'
]
```

The variable T is expanded in possitional args c and d. If we quote the variable then we get a different result:

```bash
$ export T="c d"
$ node -e 'console.log(JSON.stringify(process.env.T), process.argv)' a b "$T"
"c d" [
  '/home/ec2-user/.nvm/versions/node/v17.0.1/bin/node',
  'a',
  'b',
  'c d'
]
```

However, in the shell-quote library the situation is different. Consider the following example:

```js
const shellQuote = require('shell-quote')

console.log(shellQuote.parse('test a b $T', { T: 'c d' }))
```

We get the following result:

```js
[ 'test', 'a', 'b', 'c d' ]
```

Notice that the behaviour is different as in the variable T does not get expanded. In other words, the behaviour is similar to variable T beign quoted, i.e:

```js
const shellQuote = require('shell-quote')

console.log(shellQuote.parse('test a b "$T"', { T: 'c d' })) // [ 'test', 'a', 'b', 'c d' ]
```

If shell-quote is used to parse string which are used to spown a process with environment variable that are not quoted, it will result in a completely different behaviour than the one expected from the standard shell.

**Comment by ljharb** [comment]
interesting - this makes sense, that the presence of quotes should matter.

Are there existing tests that would break with this change?

**Comment by pdpwebsecurify** [comment]
I have no idea. I was planning to use this library but it seems that at the moment it will not be a good fit due to this particular bug.

**Comment by ljharb** [comment]
I'd love to review a PR that fixes it and adds tests :-)

**Comment by ljharb** [comment]
I've pushed up a branch with failing tests, and found the place where the unquoted env var is retrieved.

The difficulty is essentially that the current approach inlines env vars amidst parsing - however, an unquoted env var needs to be treated as part of the parse, and at this point in the implementation, we've already tokenized the input - and it's tricky to insert more tokens into the list.

**Comment by drok** [comment]
Isn't this a matter of porting an existing shell's grammar parser to js?

Here are the interesting bits from bash:

https://git.savannah.gnu.org/cgit/bash.git/tree/parse.y (Yacc grammar for bash)
https://git.savannah.gnu.org/cgit/bash.git/tree/subst.c (The part of the shell that does parameter, command, arithmetic, and globbing substitutions)

These two files account for probably 60% of what bash is.

Supporting `${T}` is not that far off from `$T`, and you'd have to support brace expansion for cases like `abc${T}def`. Then, what about `${T[1]}` or `$1` or `$PID` or `$((T*2))` or `$(echo $T | sed s/-/_/g)` ?

Are there practical applications to handling shell variable expansion in a javascript program? Can we really take shell variables out of the shell?

**Comment by ljharb** [comment]
There are many more shells than bash, and porting a parser would be extremely brittle since that can change drastically across even a single shell's major versions.

**Comment by drok** [comment]
> There are many more shells than bash, and porting a parser would be extremely brittle since that can change drastically across even a single shell's major versions.

I whole heartedly agree, and I think it's more sensible to remove variable expansion from the library, rather than attempt to fix it. I'm leaning towards removing this feature from my fork.

**Comment by ljharb** [comment]
We won’t ever be removing any features; breaking changes are toxic to the ecosystem.

**Comment by ljharb** [comment]
I've added an opt-in fix for this in b9545b3.

The new `splitUnquoted` option makes `parse` perform shell-style field splitting on unquoted variable expansions:

```js
parse('a b $T', { T: 'c d' });                          // ['a', 'b', 'c d']   (unchanged default)
parse('a b $T', { T: 'c d' }, { splitUnquoted: true }); // ['a', 'b', 'c', 'd']
```

`true` uses the default IFS (space, tab, newline). You can also pass a string to use a custom IFS, and it follows the real field-splitting rules: whitespace collapses and is trimmed at the ends, non-whitespace delimiters produce empty fields, and the two mix the way a shell does.

```js
parse('$V', { V: 'a::b' }, { splitUnquoted: ':' });   // ['a', '', 'b']
parse('$V', { V: 'a : b' }, { splitUnquoted: ' :' }); // ['a', 'b']
```

A quoted expansion (`"$T"`) is never split, matching the shell.

I made it opt-in rather than changing the default, because flipping the default would change the output of existing inputs, and that kind of break is something I try hard to avoid. The default behavior is byte-for-byte unchanged.

Scope note: this is field splitting only. It does not add globbing, and it is not a full shell (the existing parse docs describe the parameter-expansion subset that is supported).

The behavior was verified by differential testing against bash across tens of thousands of random IFS, value, and adjacency combinations.

Since this is opt-in, the default is still unsplit, so I'll leave this open as the tracker for potentially making it the default (opt-out) in a future semver-major, if we ever have one.

**Comment by volkkov** [comment]
Good catch. This is related to the RFC change in how trailing slashes are normalized. PR incoming.

**Comment by ljharb** [comment]
@volkkov what RFC?
